### PR TITLE
Add --diff support for Patches (fix #508)

### LINKED ISSUE #508: Support --diff for .atomic rewrites
With #503 we support escape hatch for rewrites. We should also support it through --diff

### PR DESCRIPTION
**Pull request by MasseGuillaume**

### CODE REVIEW

**Comment by MasseGuillaume** [comment]
@olafurpg ready for round 1.

File: scalafix-tests/input/src/main/scala/test/escapeHatch/EscapeHatchPatches.scala [hunk 1]
```diff
@@ -1,28 +1,7 @@
 /*
-rule = [RemoveUnusedImports, ExplicitResultTypes]
-ExplicitResultTypes.memberKind = [Def]
-ExplicitResultTypes.memberVisibility = [Public]
-unsafeShortenNames = true
-ExplicitResultTypes.unsafeShortenNames = true
+rule = RemoveUnusedImports
```
Comment (MasseGuillaume) [inline scalafix-tests/input/src/main/scala/test/escapeHatch/EscapeHatchPatches.scala]: ops, reverting

File: scalafix-tests/unit/src/test/scala/scalafix/tests/cli/CliGitDiffTests.scala [hunk 2]
```diff
@@ -330,6 +394,7 @@ class CliGitDiffTests() extends FunSuite with DiffAssertions {
         )
       )
       val output = new String(baos.toByteArray(), StandardCharsets.UTF_8)
+      assert(exit.isOk == ok, output)
```
Comment (olafurpg) [inline scalafix-tests/unit/src/test/scala/scalafix/tests/cli/CliGitDiffTests.scala]: > `s"Expected OK exit status, obtained $exit. Details $output"`

File: scalafix-core/shared/src/main/scala/scalafix/rule/RuleCtx.scala [hunk 3]
```diff
@@ -60,7 +60,14 @@ trait RuleCtx extends PatchOps {
...
+    apply(tree, ScalafixConfig.default, None)
+
   private[scalafix] def apply(tree: Tree, config: ScalafixConfig): RuleCtx =
-    RuleCtxImpl(tree, config)
+    apply(tree, config, None)
+
+  private[scalafix] def apply(
+      tree: Tree,
+      config: ScalafixConfig,
+      isEnabledByGitDiff: Option[Position => Boolean] = None): RuleCtx =
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/rule/RuleCtx.scala]: Alternatively, how about passing in `diffDisable: DiffDisable` with `DiffDisable.empty` as default?

File: scalafix-tests/unit/src/test/scala/scalafix/tests/cli/CliGitDiffTests.scala [hunk 4]
```diff
@@ -224,8 +225,71 @@ class CliGitDiffTests() extends FunSuite with DiffAssertions {
     assert(obtained2.startsWith(expected2))
   }
 
+  gitTest("works on Patch") { (fs, git, cli) =>
+    val code = "code.scala"
+    def show(): Unit = {
```
Comment (olafurpg) [inline scalafix-tests/unit/src/test/scala/scalafix/tests/cli/CliGitDiffTests.scala]: Is this used somewhere?

File: scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala [hunk 5]
```diff
@@ -69,34 +69,53 @@ class EscapeHatch(
...
       case AtomicPatch(underlying) =>
-        val hasEscape = {
+        val hasDisabledPatch = {
           val patches = Patch.treePatchApply(underlying)(ctx, index)
           patches.exists { tp =>
-            val (isPatchEnabled, culprit) =
-              isEnabled(name.toString, tp.tok.pos.start)
-            culprit.foreach(escape => usedEscapes += escape.offset)
-            !isPatchEnabled
+            disabledByGitDiff(tp.tok.pos) ||
```
Comment (olafurpg) [inline scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala]: What happens about "unused // scalafix:ok" warnings with this change?

Comment (MasseGuillaume) [reply scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala]: yup, let's evaluate everything and not use a shortcut operator.

**Comment by MasseGuillaume** [comment]
@olafurpg ready for round 2.

**Comment by olafurpg** [comment]
Steps to reproduce

```
git clone https://github.com/olafurpg/inThisBuild
cd inThisBuild
git checkout scalafix-diff2
sbt compile
scalafix --diff-base scalafix-diff src/main/scala/a/A.scala
```
The last command makes no changes. The diff is
```diff
diff --git a/src/main/scala/a/A.scala b/src/main/scala/a/A.scala
index ae1df34..6af6e81 100644
--- a/src/main/scala/a/A.scala
+++ b/src/main/scala/a/A.scala
@@ -1,4 +1,5 @@
 object A {
     def complex = List(1)
+    def complexAgain = List(1)
     println("")
 }

```
so I expected `complexAgain` to get an explicit result type while `complex` to remain unchanged.

Skipping --diff-base makes scalafix insert explicit types for both complex and complexAgain

**Comment by MasseGuillaume** [comment]
I will add more test for semantic rules.

**Comment by MasseGuillaume** [comment]
@olafurpg it works for me:

```
diff --git a/src/main/scala/a.scala b/src/main/scala/a.scala
index 621e832..46bc1e7 100644
--- a/src/main/scala/a.scala
+++ b/src/main/scala/a.scala
@@ -1,4 +1,5 @@
 object A {
   def complex = List(1)
+  def complexAgain = List(2)
   println("")
 }
\ No newline at end of file
```

```
~/scalafix/scalafix-nightly --diff --stdout                                                                    ~/doit
Running ExplicitResultTypes
object A {
  def complex = List(1)
  def complexAgain: _root_.scala.collection.immutable.List[_root_.scala.Int] = List(2)
  println("")
}Completed Running ExplicitResultTypes
```

**Comment by MasseGuillaume** [comment]
I reset the commit back since there is a lot of breaking changes in SemanticRuleSuite. This work is available on this branch: https://github.com/MasseGuillaume/scalafix/tree/feature/diff-test

**Comment by MasseGuillaume** [comment]
@olafurpg ready for review.

**Pull request merged**